app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php

```diff
@@ -0,0 +1,506 @@
+    }
+
+    /**
+     * @param array<Lead>    $leads
+     * @param array<Company> $companies
+     *
+     * @return array<int,array<mixed>>
+     */
+    private function setDeleteFields(array $leads, array $companies, LeadField $field): array
+    {
+        return [
+            $this->setLeadsCompaniesFieldNull($leads, $field),
+            $this->setLeadsCompaniesFieldNull($companies, $field),
+        ];
+    }
+
+    /**
+     * @param array<Lead|Company> $leadsCompanies
+     *
+     * @return array<mixed>
+     */
+    private function setLeadsCompaniesFieldNull(array $leadsCompanies, LeadField $field): array
+    {
+        foreach ($leadsCompanies as $key => $leadCompany) {
+            if (!method_exists($leadCompany, 'addUpdatedField') || !method_exists($leadCompany, 'getField')) {
+                continue;
+            }
+            if ($leadCompany instanceof Lead && 'lead' === $field->getObject()) {
+                $leadField = $leadCompany->getField($field->getAlias());
+                if (false !== $leadField) {
+                    $leadCompany->addUpdatedField($field->getAlias(), null);
+                    $leadsCompanies[$key] = $leadCompany;
+                    continue;
+                }
+            }
+            if ($leadCompany instanceof Company && 'company' === $field->getObject()) {
+                $fields    = $leadCompany->getFields();
+                $leadField = $leadCompany->getField($field->getAlias());
+
+                if (false !== $leadField) {
+                    $leadCompany->addUpdatedField($field->getAlias(), null);
+                    $leadsCompanies[$key] = $leadCompany;
+                    continue;
+                }
+
+                $alias = self::COMPANY_FIELDS_TO_COLUMNS[$field->getAlias()] ?? $field->getAlias();
+
+                $leadFieldValue = $leadCompany->getFieldValue($alias);
+                if (property_exists($leadCompany, $alias) && null !== $leadFieldValue) {
+                    $leadFieldValue = $leadCompany->{'get'.ucfirst($alias)}();
+                    if (null !== $leadFieldValue) {
+                        $leadCompany->{'set'.ucfirst($alias)}(null);
+                    }
+                    continue;
+                }
+            }
+        }
+
+        return $leadsCompanies;
+    }
+
+    /**
+     * @param array<Lead>    $leads
+     * @param array<Company> $companies
+     *
+     * @return array<int,array<mixed>>
+     */
+    private function setHashFields(array $leads, array $companies, LeadField $field, bool $pseudonymize): array
+    {
+        return [
+            $this->setHashes($leads, $field, $pseudonymize),
+            $this->setHashes($companies, $field, $pseudonymize),
+        ];
+    }
+
+    /**
+     * @param array<Lead>|array<Company> $leadsCompanies
+     *
+     * @return array<mixed>
+     */
+    private function setHashes(array $leadsCompanies, LeadField $field, bool $pseudonymize): array
```

Comment:
Consider delegating this logic to a service class to improve the separation of concerns.

Reply:
done